syntax = "proto3";

package maelstrom_client_base.items;

message StartRequest {
    optional string broker_addr = 1;
    bytes project_dir = 2;
    bytes state_dir = 3;
    bytes cache_dir = 4;
    uint64 cache_size = 5;
    uint64 inline_limit = 6;
    uint32 slots = 7;
    bytes container_image_depot_dir = 8;
}

message TarLayer {
    string path = 1;
}

message PrefixOptions {
    optional string strip_prefix = 1;
    optional string prepend_prefix = 2;
    bool canonicalize = 3;
    bool follow_symlinks = 4;
}

message GlobLayer {
    string glob = 1;
    PrefixOptions prefix_options = 2;
}

message PathsLayer {
    repeated string paths = 1;
    PrefixOptions prefix_options = 2;
}

message StubsLayer {
    repeated string stubs = 1;
}

message SymlinkSpec {
    string link = 1;
    string target = 2;
}

message SymlinksLayer {
    repeated SymlinkSpec symlinks = 1;
}

message AddLayerRequest {
    oneof Layer {
        TarLayer tar = 1;
        GlobLayer glob = 2;
        PathsLayer paths = 3;
        StubsLayer stubs = 4;
        SymlinksLayer symlinks = 5;
    }
}

enum ArtifactType {
    Tar = 0;
    Manifest = 1;
}

message LayerSpec {
    bytes digest = 1;
    ArtifactType type = 2;
}

enum JobDevice {
    Full = 0;
    Fuse = 1;
    Null = 2;
    Random = 3;
    Tty = 4;
    Urandom = 5;
    Zero = 6;
}

enum JobMountFsType {
    Proc = 0;
    Tmp = 1;
    Sys = 2;
}

message JobMount {
    JobMountFsType fs_type = 1;
    string mount_point = 2;
}

message EnvironmentSpec {
    map<string, string> vars = 1;
    bool extend = 2;
}

message ImageSpec {
    string name = 1;
    string tag = 2;
    bool use_layers = 3;
    bool use_environment = 4;
    bool use_working_directory = 5;
}

message JobSpec {
    string program = 1;
    repeated string arguments = 2;
    optional ImageSpec image = 3;
    repeated EnvironmentSpec environment = 4;
    repeated LayerSpec layers = 6;
    repeated JobDevice devices = 8;
    repeated JobMount mounts = 9;
    bool enable_loopback = 10;
    bool enable_writable_file_system = 11;
    optional string working_directory = 12;
    uint32 user = 13;
    uint32 group = 14;
    optional uint32 timeout = 15;
    optional Duration estimated_duration = 16;
}

message RunJobRequest {
    JobSpec spec = 1;
}

message GetContainerImageRequest {
    string name = 1;
    string tag = 2;
}

message Error {
    string message_ = 1;
}

message AddLayerResponse {
    LayerSpec spec = 1;
}

message OciConfig {
    optional string user = 1;
    repeated string exposed_ports = 2;
    repeated string env = 3;
    repeated string entrypoint = 4;
    repeated string cmd = 5;
    repeated string volumes = 6;
    optional string working_dir = 7;
    map<string, string> labels = 8;
    optional string stop_signal = 9;
}

message OciRootFs {
    string type = 1;
    repeated string diff_ids = 2;
}

message OciImageConfiguration {
    optional string created = 1;
    optional string author = 2;
    optional string architecture = 3;
    optional string os = 4;
    optional string os_version = 5;
    repeated string os_features = 6;
    optional string variant = 7;
    optional OciConfig config = 8;
    OciRootFs rootfs = 9;
}

message ContainerImage {
    uint32 version = 1;
    string name = 2;
    string digest = 3;
    OciImageConfiguration config = 4;
    repeated bytes layers = 5;
}

message GetContainerImageResponse {
    ContainerImage image = 1;
}

message JobError {
    oneof kind {
        string execution = 1;
        string system = 2;
    }
}

message JobOutputResultTruncated {
    bytes first = 1;
    uint64 truncated = 2;
}

message JobOutputResult {
    oneof result {
        Void none = 1;
        bytes inline = 2;
        JobOutputResultTruncated truncated = 3;
    }
}

message Duration {
    uint64 seconds = 1;
    uint32 nano_seconds = 2;
}

message JobEffects {
    JobOutputResult stdout = 1;
    JobOutputResult stderr = 2;
    Duration duration = 3;
}

message JobCompleted {
    oneof status {
        uint32 exited = 1;
        uint32 signaled = 2;
    }
    JobEffects effects = 3;
}

message JobOutcome {
    oneof outcome {
        JobCompleted completed = 1;
        JobEffects timed_out = 2;
    }
}

message JobOutcomeResult {
    oneof result {
        JobError error = 1;
        JobOutcome outcome = 2;
    }
}

message RunJobResponse {
    uint32 client_job_id = 1;
    JobOutcomeResult result = 2;
}

message Void {}

message JobStateCounts {
    uint64 waiting_for_artifacts = 1;
    uint64 pending = 2;
    uint64 running = 3;
    uint64 complete = 4;
}

message RemoteProgress {
    string name = 1;
    uint64 size = 2;
    uint64 progress = 3;
}

message IntrospectResponse {
    repeated RemoteProgress artifact_uploads = 1;
    repeated RemoteProgress image_downloads = 2;
    JobStateCounts job_state_counts = 3;
}

service ClientProcess {
    rpc Start(StartRequest) returns (Void);
    rpc AddLayer(AddLayerRequest) returns (AddLayerResponse);
    rpc GetContainerImage(GetContainerImageRequest) returns (GetContainerImageResponse);
    rpc RunJob(RunJobRequest) returns (RunJobResponse);
    rpc Introspect(Void) returns (IntrospectResponse);
}
